File: RSASign_Sample_Objc/RSASign_Sample_Objc/RSASignViewController.m
//
//  RSASignViewController.m
//  RSASign_Sample_Objc
//
//  Created by Yunju on 2020/11/20.
//

#import "RSASignViewController.h"
#import "RSASign.h"

#import <CommonCrypto/CommonDigest.h>
#import <CommonCrypto/CommonHMAC.h>
#import <CommonCrypto/CommonCryptor.h>

#define RESULT_CODE_SUCCESS  @"0000"

@interface RSASignViewController () <UITextViewDelegate> {
    UITapGestureRecognizer *bgTap;
}

@end

@implementation RSASignViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    bgTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(writeFinished)];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleKeyboardWillShowNote:) name:UIKeyboardWillShowNotification object:self.view.window];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleKeyboardWillHideNote:) name:UIKeyboardWillHideNotification object:self.view.window];
}

#pragma mark - Commons
- (void) writeFinished {
    [self.view endEditing:YES];
}

-(void)showResult:(NSString *)title message:(NSString *)message
{
    UIAlertController * alert=   [UIAlertController
                                  alertControllerWithTitle:title
                                  message:message
                                  preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* ok = [UIAlertAction
                         actionWithTitle:@"OK"
                         style:UIAlertActionStyleDefault
                         handler:^(UIAlertAction * action)
                         {
                             [alert dismissViewControllerAnimated:YES completion:nil];
                         }];
    
    [alert addAction:ok];
    
    [self presentViewController:alert animated:YES completion:nil];
}

-(NSString*)jsonStringPrint : (NSDictionary *)dic
{
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dic
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:nil];
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

-(NSData *) hexToByteArray:(NSString *)hex
{
    if (hex.length == 0) { return nil; }
    
    static const unsigned char HexDecodeChars[] =
    {
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 1, //49
        2, 3, 4, 5, 6, 7, 8, 9, 0, 0, //59
        0, 0, 0, 0, 0, 10, 11, 12, 13, 14,
        15, 0, 0, 0, 0, 0, 0, 0, 0, 0, //79
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 10, 11, 12, //99
        13, 14, 15
    };
    
    // convert data(NSString) to CString
    const char *source = [hex cStringUsingEncoding:NSUTF8StringEncoding];
    // malloc buffer
    unsigned char *buffer;
    NSUInteger length = strlen(source) / 2;
    buffer = malloc(length);
    for (NSUInteger index = 0; index < length; index++) {
        buffer[index] = (unsigned char)(HexDecodeChars[(size_t)source[index * 2]] << 4) + (HexDecodeChars[(size_t)source[index * 2 + 1]]);
    }
    
    // init result NSData
    NSData *result = [NSData dataWithBytes:buffer length:length];
    free(buffer);
    source = nil;
    
    return result;
}

-(NSString*)dataToHexString:(NSData*)data {
    size_t size = [data length];
    const uint8_t * const randomBytes = [data bytes];
    NSMutableString *hexStr;
    hexStr = [[NSMutableString alloc] initWithCapacity:size];
    for(NSInteger index = 0; index < data.length; index++)
    {
        [hexStr appendFormat: @"%02x", randomBytes[index]];
    }
    return hexStr;
}

#pragma mark - UITextViewDelegate
- (BOOL)textViewShouldEndEditing:(UITextView *)textView {
    [textView resignFirstResponder];
    return YES;
}

- (void)textViewDidEndEditing:(UITextView *)textView {
}

- (void)textViewDidChange:(UITextView *)textView {
}

#pragma mark - keyboard actions
- (void)handleKeyboardWillShowNote:(NSNotification *)notification
{
    [self.view addGestureRecognizer:bgTap];
}

- (void)handleKeyboardWillHideNote:(NSNotification *)notification
{
    [self.view removeGestureRecognizer:bgTap];
}

#pragma mark - Buton Actions
- (IBAction)actionBtnVersion:(id)sender {
    NSString *title = @"Get library version";
    __block NSString *msg = nil;
    
    // Sync
    NSDictionary *resDic = [RSASign getVersion];
    msg = [self jsonStringPrint:resDic];
    [self showResult:title message:msg];
    
    // Async
//    [RSASign getVersion:^(BOOL success, NSDictionary *result) {
//        msg = [self jsonStringPrint:result];
//        [self showResult:title message:msg];
//    }];
}

- (IBAction)actionBtnGenKey:(id)sender {
    NSString *title = @"Generate RSA key";
    __block NSString *msg = nil;
    
    [self.textSignature setText:@""];
    
    // Sync
    __block NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:[RSASign generateKey]];
    if ([[resDic objectForKey:@"resultCode"] isEqualToString:RESULT_CODE_SUCCESS]) {
        NSString *pubKey = [self dataToHexString:[resDic objectForKey:@"publicKey"]];
        [resDic setObject:pubKey forKey:@"publicKey"];
    }
    msg = [self jsonStringPrint:resDic];
    [self showResult:title message:msg];
    
    
    // Async
//    [RSASign generateKey:^(BOOL success, NSDictionary *result) {
//        if (success) {
//            NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:result];
//            NSString *pubKey = [self dataToHexString:[result objectForKey:@"publicKey"]];
//            [resDic setObject:pubKey forKey:@"publicKey"];
//        }
//        msg = [self jsonStringPrint:resDic];
//        [self showResult:title message:msg];
//    }];
}

- (IBAction)actionBtnGetPubKey:(id)sender {
    NSString *title = @"Get public key";
    __block NSString *msg = nil;
    
    // Sync
    __block NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:[RSASign getPublicKey]];
    if ([[resDic objectForKey:@"resultCode"] isEqualToString:RESULT_CODE_SUCCESS]) {
        NSString *pubKey = [self dataToHexString:[resDic objectForKey:@"publicKey"]];
        [resDic setObject:pubKey forKey:@"publicKey"];
    }
    msg = [self jsonStringPrint:resDic];
    [self showResult:title message:msg];
    
    // Async
//    [RSASign getPublicKey:^(BOOL success, NSDictionary *result) {
//        if (success) {
//            NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:result];
//            NSString *pubKey = [self dataToHexString:[resDic objectForKey:@"publicKey"]];
//            [resDic setObject:pubKey forKey:@"publicKey"];
//        }
//        msg = [self jsonStringPrint:resDic];
//        [self showResult:title message:msg];
//    }];
}

- (IBAction)actionBtnCreateSign:(id)sender {
    NSString *title = @"Create Signature";
    __block NSString *msg = nil;
    
    NSData *signData = [@"original data" dataUsingEncoding:NSUTF8StringEncoding];
    
    // Sync
    __block NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:[RSASign createSignature:signData]];
    if ([[resDic objectForKey:@"resultCode"] isEqualToString:RESULT_CODE_SUCCESS]) {
        NSString *signature = [self dataToHexString:[resDic objectForKey:@"signature"]];
        [resDic setObject:signature forKey:@"signature"];
        
        [self.textSignature setText:signature];
    }
    msg = [self jsonStringPrint:resDic];
    [self showResult:title message:msg];
        
    // Async
//    [RSASign createSignature:signData callback:^(BOOL success, NSDictionary *result) {
//        if (success) {
//            NSString *signature = [self dataToHexString:[resDic objectForKey:@"signature"]];
//            [resDic setObject:signature forKey:@"signature"];
//
//            [self.textSignature setText:signature];
//        }
//        msg = [self jsonStringPrint:result];
//        [self showResult:title message:msg];
//    }];
}

- (IBAction)actionBtnVerifySign:(id)sender {
    NSString *title = @"Verify Signature";
    __block NSString *msg = nil;
    
    NSData *signData = [@"original data" dataUsingEncoding:NSUTF8StringEncoding];
    NSData *signature = [self hexToByteArray:self.textSignature.text];
//    NSData *signature = [self.textSignature.text dataUsingEncoding:NSUTF8StringEncoding];
        
    // Sync
    __block NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:[RSASign verifySignature:signData signature:signature]];
    msg = [self jsonStringPrint:resDic];
    [self showResult:title message:msg];
    
    // Async
//    [RSASign verifySignature:signData signature:signature callback:^(BOOL success, NSDictionary *result) {
//        msg = [self jsonStringPrint:resDic];
//        [self showResult:title message:msg];
//    }];
}

- (IBAction)actionBtnDelKey:(id)sender {
    NSString *title = @"Delete RSA Key";
    __block NSString *msg = nil;
    
    [self.textSignature setText:@""];
    
    // Sync
    __block NSMutableDictionary *resDic = [NSMutableDictionary dictionaryWithDictionary:[RSASign deleteKey]];
    msg = [self jsonStringPrint:resDic];
    [self showResult:title message:msg];
    
    // Async
//    [RSASign deleteKey:^(BOOL success, NSDictionary *result) {
//        msg = [self jsonStringPrint:resDic];
//        [self showResult:title message:msg];
//    }];
}

@end

File: RSASign_Sample_Swift/RSASign_Sample_Swift/RSASignViewController.swift
//
//  RSASignViewController.swift
//  MobileRSign_iOS_Swift
//
//  Created by Yunju on 03/09/2019.
//  Copyright © 2019 ATON. All rights reserved.
//

import UIKit
import CommonCrypto.CommonDigest
import CommonCrypto.CommonHMAC
import CommonCrypto.CommonCryptor

let RESULT_CODE_SUCCESS =     "0000"

class RSASignViewController: UIViewController, UITextViewDelegate {
    
    @IBOutlet weak var textSignature: UITextView!
    
    private var bgTap: UITapGestureRecognizer?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        bgTap = UITapGestureRecognizer.init(target: self, action: #selector(writeFinished))
        
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide), name: UIResponder.keyboardWillHideNotification, object: nil)
    }
    
    // MARK: Common Functions
    @objc func writeFinished()
    {
        self.view.endEditing(true)
    }
    
    func showResult (title: String, message: String)
    {
        let alert = UIAlertController.init(title: title, message: message, preferredStyle: UIAlertController.Style.alert)
        alert.addAction(UIAlertAction.init(title: "OK", style: UIAlertAction.Style.default, handler: {
            action in
            alert.dismiss(animated: true, completion: nil)
            
        }))
        
        self.present(alert, animated: true, completion: nil)
    }
    
    func jsonStringPrint (dic: [AnyHashable : Any]) -> String
    {
        let jsonData: Data! = try? JSONSerialization.data(withJSONObject: dic, options: JSONSerialization.WritingOptions.prettyPrinted)
        let jsonString: String! = String(data: jsonData, encoding: String.Encoding.utf8)
        
        return jsonString ?? ""
    }
    
    func hexToBytes(_ string: String) -> [UInt8]? {
        let length = string.count
        if length & 1 != 0 {
            return nil
        }
        var bytes = [UInt8]()
        bytes.reserveCapacity(length/2)
        var index = string.startIndex
        for _ in 0..<length/2 {
            let nextIndex = string.index(index, offsetBy: 2)
            if let b = UInt8(string[index..<nextIndex], radix: 16) {
                bytes.append(b)
            } else {
                return nil
            }
            index = nextIndex
        }
        return bytes
    }
    
    // MARK: UITextViewDelegate
    func textViewShouldEndEditing(_ textView: UITextView) -> Bool {
        textView.resignFirstResponder()
        return true
    }
    
    func textViewDidEndEditing(_ textView: UITextView) {
    }
    
    func textViewDidChange(_ textView: UITextView) {
    }
    
    // MARK: Actions
    @IBAction func actionBtnVersion(_ sender: Any) {
        let title: String! = "Get library version"
        var msg: String! = ""
        
        // Sync
        let resDic: Dictionary! = RSASign.getVersion()
        msg = self.jsonStringPrint(dic: resDic)
        self.showResult(title: title, message: msg)
        
        // Async
//        RSASign.getVersion({ (success: Bool, result: Dictionary) in
//            msg = self.jsonStringPrint(dic: result)
//            self.showResult(title: title, message: msg)
//        })
    }
    
    @IBAction func actionBtnGenKey(_ sender: Any) {
        let title: String! = "Generate RSA key"
        var msg: String! = ""
        
        self.textSignature.text = ""
        
        // Sync
        var resDic: Dictionary! = RSASign.generateKey()
        if (resDic["resultCode"] as! String? == RESULT_CODE_SUCCESS) {
            let pubKey: Data! = resDic["publicKey"] as? Data
            resDic["publicKey"] = pubKey.hexEncodedString()
        }
        msg = self.jsonStringPrint(dic: resDic);
        self.showResult(title: title, message: msg)
        
        // Async
//        RSASign.generateKey({ (success: Bool, result: Dictionary) in
//            if success == true {
//                var resDic: Dictionary = result;
//                let pubKey: Data! = resDic["publicKey"] as? Data
//                resDic["publicKey"] = pubKey.hexEncodedString()
//            }
//            msg = self.jsonStringPrint(dic: resDic);
//            self.showResult(title: title, message: msg)
//        })
    }
    
    @IBAction func actionBtnGetPubKey(_ sender: Any) {
        let title: String! = "Get public key"
        var msg: String! = ""

        // Sync
        var resDic: Dictionary! = RSASign.getPublicKey()
        if (resDic["resultCode"] as! String? == RESULT_CODE_SUCCESS) {
            let pubKey: Data! = resDic["publicKey"] as? Data
            resDic["publicKey"] = pubKey.hexEncodedString()
        }
        msg = self.jsonStringPrint(dic: resDic);
        self.showResult(title: title, message: msg)
        
        // Async
//        RSASign.getPublicKey({ (success: Bool, result: Dictionary) in
//            if success == true {
//                var resDic: Dictionary = result;
//                let pubKey: Data! = resDic["publicKey"] as? Data
//                resDic["publicKey"] = pubKey.hexEncodedString()
//            }
//            msg = self.jsonStringPrint(dic: resDic);
//            self.showResult(title: title, message: msg)
//        })
    }
    
    @IBAction func actionBtnCreateSign(_ sender: Any) {
        let title: String! = "Create Signature"
        var msg: String! = ""
        
        let signData: Data! = "original data".data(using: .utf8)

        // Sync
        var resDic: Dictionary! = RSASign.createSignature(signData)
        if resDic["resultCode"] as! String? == RESULT_CODE_SUCCESS {
            let signature: Data! = resDic["signature"] as? Data
            resDic["signature"] = signature.hexEncodedString()
            
            self.textSignature.text = signature.hexEncodedString()
        }
        msg = self.jsonStringPrint(dic: resDic);
        self.showResult(title: title, message: msg)
            
        // Async
//        RSASign.createSignature(signData, callback:{ (success: Bool, result: Dictionary) in
//            if success == true {
//                let signature: Data! = resDic["signature"] as? Data
//                resDic["signature"] = signature.hexEncodedString()
//
//                self.textSignature.text = signature.hexEncodedString()
//            }
//            msg = self.jsonStringPrint(dic: result);
//            self.showResult(title: title, message: msg)
//        })
    }
    
    @IBAction func actionBtnVerifySign(_ sender: Any) {
        let title: String! = "Verify Signature"
        var msg: String! = ""
        
        let signData: Data! = "original data".data(using: .utf8)
        let signature: [UInt8]! = self.hexToBytes(self.textSignature.text)
        let signatureData: NSData! = NSData(bytes: signature, length: signature.count)
        
        // Sync
        let resDic: Dictionary! = RSASign.verifySignature(signData, signature: (signatureData as Data?))
        msg = self.jsonStringPrint(dic: resDic);
        self.showResult(title: title, message: msg)
        
        // Async
//        RSASign.verifySignature(signData, signature: (signatureData as Data?), callback:{ (success: Bool, result: Dictionary) in
//            msg = self.jsonStringPrint(dic: resDic);
//            self.showResult(title: title, message: msg)
//        })
    }
    
    @IBAction func actionBtnDelKey(_ sender: Any) {
        let title: String! = "Delete RSA key"
        var msg: String! = ""
        
        self.textSignature.text = ""
        
        // Sync
        let resDic: Dictionary! = RSASign.deleteKey()
        msg = self.jsonStringPrint(dic: resDic);
        self.showResult(title: title, message: msg)
        
        // Async
//        RSASign.deleteKey({ (success: Bool, result: Dictionary) in
//            msg = self.jsonStringPrint(dic: result);
//            self.showResult(title: title, message: msg)
//        })
    }
    
    
    // MARK: Keyboard Notification
    @objc func keyboardWillShow(notification: NSNotification) {
        view.addGestureRecognizer(bgTap!)
    }
    
    @objc func keyboardWillHide(notification: NSNotification) {
        view.removeGestureRecognizer(bgTap!)
    }
    
}
